teps/0044-composing-tasks-with-tasks.md

```diff
@@ -0,0 +1,137 @@
+---
+status: proposed
+title: Composing Tasks with Tasks
+creation-date: '2021-01-22'
+last-updated: '2021-01-22'
+authors:
+- '@bobcatfish'
+---
+
+# TEP-0044: Composing Tasks with Tasks
```

Comment:
Thank you for staring this 🙏 

I agree that we need a way to compose tasks that is more efficient than what we have today.
I feel that the problem is not on the authoring side though, but on the runtime.

Today we compose Tasks using Pipeline, which gives us a rich model (perhaps a bit verbose) to connect tasks through params, workspaces, results and more. In itself this model allows us to write highly re-usable tasks that can be combined into complex pipelines. The API is beta, and it has a reasonable level of adoption. 

On the runtime side, when we execute a pipeline against a k8s cluster, we force each task to create a new Pod, which means that today - if we use catalog tasks - we need two pods and a pvc to do something like cloning a repo and building a container image. Some very common use cases like this were covered by pipeline resources; removing them from the pictures highlights the need for alternative runtime models.

To summarise, I think the solution to the problem highlighted in this TEP will not be a new way of composing tasks in YAML, but a new way to execute them on a k8s cluster.

Reply:
"create a Pipeline that groups some Tasks such that they are run on the same pod" -- this is exactly what our teams have been asking us for. Their issue is that if we write fine-grained tasks to promote re-use in our task catalogs we incur a roughly 10 or so second overhead when running a sequence of Tasks. @afrittoli's point about doing the task/pod aggregation automatically is definitely interesting but I would be delighted to be able to do this manually as a first step.